### PR TITLE
Shortcuts to comment / uncomment lines

### PR DESCRIPTION
**Pull request by yescorp**

### CODE REVIEW

File: lib/src/code_field/code_controller.dart [hunk 1]
```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
...
+      // to avoid including the next line if `\n` is selected
+      selection.isCollapsed ? selection.end : selection.end - 1,
+      placeHiddenRanges: TextAffinity.downstream,
+    );
+
+    // index of the first line to modify
+    final firstLineIndex =
+        _code.lines.characterIndexToLineIndex(selectionStart);
+    // index of the last line to modify
+    final lastLineIndex = _code.lines.characterIndexToLineIndex(selectionEnd);
```
Comment (alexeyinkin) [inline lib/src/code_field/code_controller.dart:490]: Extract to `TextRange getSelectedLineRange()`, reuse it in `modifySelectedLines`.

`TextRange.start` should be the first selected line. `TextRange.end` should be the last selected line + 1.

File: lib/src/code_field/code_controller.dart [hunk 2]
```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
...
+      );
+    });
+  }
+
+  /// Utility method to divide [commentOrUncommentSelection] method into parts
+  /// [commentSymbols] - `//` or `#` indicates what to use to uncomment
+  void _uncommentSelectedLines(String commentSymbols) {
+    modifySelectedLines((line) {
+      // if line is empty skip it
+      if (RegExps.emptyLine.hasMatch(line)) {
```
Comment (alexeyinkin) [inline lib/src/code_field/code_controller.dart:457]: Make an extension method on String: `bool hasOnlyWhitespaces() => trim() == ''`

File: test/src/code_field/code_controller_comment_uncomment_test.dart [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+import 'package:highlight/languages/java.dart';
+
+void main() {
+  group('Selection is collapsed, language: java', () {
+    test(
+        'WHEN selection is collapsed '
+        'SHOULD comment selected line if it is not a comment '
+        'and uncomment otherwise '
+        'AND select whole lines', () {
+      final examples = [
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_comment_uncomment_test.dart:15]: More ideas for tests:
- ` // A` to comment
- Empty lines to comment out.
- Empty lines to uncomment.
- Different indentation when commenting out.
- Different indentation when uncommenting.

File: lib/src/code_field/code_controller.dart [hunk 4]
```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
...
+        startOfCommentSymbols!,
+        startOfCommentSymbols,
+        '$commentSymbols ',
+      );
+    });
+  }
+
+  /// Utility method to divide [commentOrUncommentSelection] method into parts
+  /// [commentSymbols] - `//` or `#` indicates what to use to uncomment
+  void _uncommentSelectedLines(String commentSymbols) {
```
Comment (alexeyinkin) [inline lib/src/code_field/code_controller.dart:454]: Remove the argument, handle any supported sequences.

File: lib/src/code_field/code_controller.dart [hunk 5]
```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
...
+      // if line is empty skip it
+      if (RegExps.emptyLine.hasMatch(line)) {
+        return line;
+      }
+
+      // replace first `// ` or `# ` or `//` or `#` with empty string
+      return line.replaceFirst(
+        RegExps.getCommentPlusWhitespaceRegExp(commentSymbols),
+        '',
+      );
```
Comment (alexeyinkin) [inline lib/src/code_field/code_controller.dart:465]: I thinks this is more readable and can be better verified visually than a reg exp:
```suggestion
      for (final sequence in SingleLineComments.byMode[language] ?? []) {
        if (line.trim().startsWith(sequence)) {
          // ...
        }
      }
```

File: lib/src/code_field/code_controller.dart [hunk 6]
```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
...
+        return;
+      } else {
+        _commentSelectedLines(commentType);
+        return;
+      }
+    }
+  }
+
+  /// Utility method to divide [commentOrUncommentSelection] method into parts
+  /// [commentSymbols] - `//` or `#` indicates what to use to comment
```
Comment (alexeyinkin) [inline lib/src/code_field/code_controller.dart:427]: Delete.

File: lib/src/code_field/code_controller.dart [hunk 7]
```diff
@@ -380,6 +384,121 @@ class CodeController extends TextEditingController {
...
+
+      if (startOfCommentSymbols == null) {
+        return '$commentSymbols $line';
+      }
+
+      return line.replaceRange(
+        startOfCommentSymbols!,
+        startOfCommentSymbols,
+        '$commentSymbols ',
+      );
```
Comment (alexeyinkin) [inline lib/src/code_field/code_controller.dart:448]: Make an extension getter on String: `int get startWhitespaceCount`.
Then make a single return without `if`, less conditions is better.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: test/src/code_field/code_controller_comment_uncomment_test.dart [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+import 'package:flutter_test/flutter_test.dart';
+import 'package:highlight/languages/java.dart';
+
+void main() {
+  group('Selection is collapsed, language: java', () {
+    test(
+        'WHEN selection is collapsed '
+        'SHOULD comment selected line if it is not a comment '
+        'and uncomment otherwise '
+        'AND select whole lines', () {
```
[the remark this answers is not in the document]
Comment (yescorp) [reply test/src/code_field/code_controller_comment_uncomment_test.dart:14]: Also added php Mode to the dictionary

File: lib/src/code_field/code_controller.dart (hunk 7 above)
Comment (yescorp) [reply lib/src/code_field/code_controller.dart]: I removed this logic entirely.
Now we just add comments to the start of the line

File: lib/src/code_field/code_controller.dart [hunk 9]
```diff
@@ -431,6 +519,11 @@ class CodeController extends TextEditingController {
 
     final modifiedLinesString = modifiedLinesBuffer.toString();
 
+    final firstLineStart =
+        _code.lines.lines[selectedLinesRange.start].textRange.start;
+    final lastLineEnd =
+        _code.lines.lines[selectedLinesRange.end - 1].textRange.end;
```
Comment (yescorp) [inline lib/src/code_field/code_controller.dart:525]: Moved it here to make it more readable

File: lib/src/code_field/code_field.dart [hunk 10]
```diff
@@ -75,6 +76,16 @@ final _shortcuts = {
...
+
+  // Comment Uncomment
+  LogicalKeySet(
+    LogicalKeyboardKey.control,
+    LogicalKeyboardKey.slash,
+  ): const CommentUncommentIntent(),
+  const SingleActivator(
+    LogicalKeyboardKey.slash,
+    meta: true,
+  ): const CommentUncommentIntent(),
```
Comment (yescorp) [inline lib/src/code_field/code_field.dart:88]: Due to this issue https://github.com/flutter/flutter/issues/103964
I suggest to also replace other `LogicalKeySet`-s  that have `meta` with `SingleActivator`

File: lib/src/single_line_comments/parser/single_line_comments.dart [hunk 11]
```diff
@@ -14,6 +15,7 @@ class SingleLineComments {
     go: [_slashes],
     scala: [_slashes],
     python: [_hash],
+    php: [_slashes, _hash],
```
Comment (alexeyinkin) [inline lib/src/single_line_comments/parser/single_line_comments.dart:18]: Sort this list please.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: test/src/code_field/code_controller_comment_uncomment_test.dart [hunk 12]
```diff
@@ -4,250 +4,506 @@ import 'package:flutter/material.dart';
...
+
+# Aa
+''',
+            initialSelection: TextSelection(baseOffset: 4, extentOffset: 15),
+            expectedFullText: '''
+aA
+AA
+
+Aa
+''',
```
[the remark this answers is not in the document]
Comment (yescorp) [reply test/src/code_field/code_controller_comment_uncomment_test.dart:384]: That's expected. 
If none of the lines are uncommented (all of them are comments) we remove uncomment the lines.

And it will not cause to return false.
false is returned only if every line returns false.
If there is `a` line that contains text but is not a comment => true

File: test/src/code_field/code_controller_comment_uncomment_test.dart [hunk 13]
```diff
@@ -15,105 +15,30 @@ void main() {
           'and uncomment otherwise '
           'AND select whole lines', () {
         final examples = [
+          //
```
Comment (alexeyinkin) [inline test/src/code_field/code_controller_comment_uncomment_test.dart:18]: Add tests for invalid selection of `-1` (must do nothing).

Also looks like we now have only tests that change each line. We need at least some tests with deselected lines that do not change, although we do not need all edge cases for that.

**Review by alexeyinkin — CHANGES_REQUESTED** [review]
